### DISCUSSION TITLE
Offline mode

### DISCUSSION THREAD

**Discussion by ippocratis** [Radiokot/photoprism-android-client#16] [Ideas]
Being able to navigate a bit in the app
while it is indeed snapy I think it suffers from  one of the basic disadvantages the pwa do
There is no offline mode
Being a proper android app maybe there is a way to cache the alreafy cahed thumbnails and photos/videos for ofline view
to take it even further an option to predownload all thumbnails would also probably make sense for even faster scrolling thumbnail loads

Example app that does the same (the pre-caching). 
It is a Kotlin android client for another popular photo managment app librephotos https://github.com/savvasdalkitsis/uhuruphotos-android/commit/ebd77b0aa49c7b3181c6a0524b383a8f82f85e0a

On the other hand I'm not sure this is even doable with web service api (or maybe it is totaly indipended from the api?)

Thanks
keep up

**Comment by ippocratis** [comment]
Some more observations on this

App will continue to display some cahed thumbnails and even open some cached images as long as its kept in the android recent apps

**Comment by Radiokot** [comment]
The app has an image cache of a limited size, which behaves like a browser cache. So it reads previously loaded images from the file system if their lifetime is not expired, which is determined by the Cache-Control headers sent by the server. PhotoPrism sets the lifetime to maximum, so cached images only get removed if the cache size limit is reached and they haven't been displayed recently.

I agree that the offline mode is convenient and could be useful. 
I think the minimal "lightweight" offline can be described as follows:
- The app downloads all the gallery media data for a default search request (order:newest, public:true)
- The download/refresh is initiated manually
- If there is no downloaded data, the online paged loading is used (the way the app currently loads the data)
- When applied a custom search query, the online page loading is used
- Thumbnails and previews are cached as you view the media, the cache size is limited (the way the cache currently works)

Such strategy will improve scrolling, save some traffic and allow you to preview media you view often.

Yet I don't feel enthusiastic at all for the implementation, the main reason for which is PhotoPrism not having an API to make it right. To refresh the offline storage, you have to re-load all your library as for the first time. Such process not only takes long time and traffic, but is more prone to errors hence requires advanced resolution strategies. It is doable, and I've done this before, I just don't want to put so much effort to it right now.

**Reply by becash** [reply]
hi @Radiokot I'm traveling now, and feel that missing offline mode for your , as I considered COOL application,  is a option that is must have.  im a  backennd developer GO, and python. but no kotlin,  maybe together we we can make a simple caching for your application. I see that photoprism is written in Go, maybe I can make something api methods and you implemented them in your application?

**Reply by Radiokot** [reply]
Hi. I'm glad you enjoy the gallery, and thank you for offering help, but I don't think it's a good idea. If you need your content offline, I recommend using Autosync to set up WebDAV folder synchronization.
